src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/RazorCompletionItem.cs

```diff
@@ -2,110 +2,82 @@
-    /// <param name="insertText">Content to insert when completion item is committed</param>
+    /// <param name="displayText">The text to display in the completion list.</param>
+    /// <param name="insertText">Content to insert when completion item is committed.</param>
     /// <param name="kind">The type of completion item this is. Used for icons and resolving extra information like tooltip text.</param>
+    /// <param name="descriptionInfo">An object that provides description information for this completion item.</param>
     /// <param name="sortText">A string that is used to alphabetically sort the completion item. If omitted defaults to <paramref name="displayText"/>.</param>
     /// <param name="commitCharacters">Characters that can be used to commit the completion item.</param>
-    /// <exception cref="ArgumentNullException">Thrown if <paramref name="displayText"/> or <paramref name="insertText"/> are <c>null</c>.</exception>
     /// <param name="isSnippet">Indicates whether the completion item's <see cref="InsertText"/> is an LSP snippet or not.</param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="displayText"/> or <paramref name="insertText"/> are <see langword="null"/>.</exception>
     public RazorCompletionItem(
         string displayText,
         string insertText,
         RazorCompletionItemKind kind,
+        object? descriptionInfo = null,
         string? sortText = null,
         ImmutableArray<RazorCommitCharacter> commitCharacters = default,
         bool isSnippet = false)
     {
-        if (displayText is null)
-        {
-            throw new ArgumentNullException(nameof(displayText));
-        }
-
-        if (insertText is null)
-        {
-            throw new ArgumentNullException(nameof(insertText));
-        }
+        ArgHelper.ThrowIfNull(displayText);
+        ArgHelper.ThrowIfNull(insertText);
 
         DisplayText = displayText;
         InsertText = insertText;
         Kind = kind;
+        DescriptionInfo = descriptionInfo;
         CommitCharacters = commitCharacters.NullToEmpty();
         SortText = sortText ?? displayText;
         IsSnippet = isSnippet;
     }
 
-    public string DisplayText { get; }
-
-    public string InsertText { get; }
-
-    public bool IsSnippet { get; }
-
-    /// <summary>
-    /// A string that is used to alphabetically sort the completion item.
-    /// </summary>
-    public string SortText { get; }
-
-    public RazorCompletionItemKind Kind { get; }
-
-    public ImmutableArray<RazorCommitCharacter> CommitCharacters { get; }
-
-    public ItemCollection Items
-    {
-        get
-        {
-            if (_items is null)
-            {
-                lock (this)
-                {
-                    _items ??= new ItemCollection();
-                }
-            }
-
-            return _items;
-        }
-    }
+    public static RazorCompletionItem CreateMarkupTransition(
```

Comment:
>CreateMarkupTransition [](http://example.com/codeflow?start=38&length=22)

Nit: I personally would've preferred this added to MarkupTransitionCompletionItemProvider since that class is already specific to MarkupTransition.

Reply:
I was doing review commit-by-commit - looking at the last commit, I see we added all factory methods to the completion item itself. I personally would've added them in the corresponding provider types as those are already specific to those more specific item instances, but don't have very strong feelings either way.

Reply:
The purpose of these factory methods was to ensure that a `RazorCompletionItem` is always created correctly. To support that goal, I want to enforce that `RazorCompletionItem` can _only_ be constructed via factory methods. If the factory methods aren't on `RazorCompletionItem`, the constructor would have to be public and there wouldn't be any enforcement.

Reply:
> I personally would've preferred this added to MarkupTransitionCompletionItemProvider since that class is already specific to MarkupTransition.

FYI that I appreciate you explaining your thinking and that's a good place to start logically. For me, when I look at `RazorCompletionItem`, I see the logic for MarkupTransition already split across other related areas, such as `RazorCompletionItemResolver`.